doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,229 @@
+> - `SELECT`, `INSERT`, `UPDATE`, and `DELETE` statements (but not `TAIL`)
+> - Materialize initiated acknowledgements of upstream sources (*e.g., the commit of an offset by a Kafka source, the acknowledge of an LSN by a PostgresSQL source, etcetera*)
+
+This document does not take any opinion on whether the changes proposed should be the default behavior or an opt-in
+behavior via some configuration.
+
+Time units of milliseconds are used below for convenience to reflect current implementations. However, this document
+does not take any opinion on what time units should be used for the design proposed.
+
+## Goals
+
+The goal of this design document is to describe what changes are needed to provide Strict Serializability to the
+operations mentioned in correctness property 1.
+
+## Non-Goals
+
+Sinks are left out of this document and not considered for consistency.
+
+`TAIL` is left out of this document and not considered for consistency.
+
+`AS OF` allows clients to specify the timestamp of a query. Queries with `AS OF` do not constrain the timestamp
+selection of other queries. These properties allow clients to circumvent our consistency guarantees, so `AS OF` is left
+out of this document.
+
+## Description
+
+Strict Serializability (in simplified terms) means that you can assign a total order to all transactions in the system,
+and that the total order is constrained by the real time occurrences of those transactions. In other words, if
+transaction A finishes in real time before transaction B starts in real time, then transaction A MUST be ordered before
+transaction B in the total order. Concurrent transaction can be ordered arbitrarily. For a more in depth discussion
+please
+read [https://jepsen.io/consistency/models/strict-serializable](https://jepsen.io/consistency/models/strict-serializable)
+and chapters 7 and 9 of [Designing Data-Intensive Applications](https://dataintensive.net/).
+
+Materialize can satisfy the constraints of Strict Serializability if all reads and writes occur at specific timestamps,
+and these timestamps are non-decreasing in real-time order, with strict increases for writes following reads. Each
+timestamp is assigned to an event at some real-time moment within the event's bounds (its start and stop). Each
+timestamp transition is made durable before any subsequent response is issued.
+
+We can then use these timestamp to form a total order constrained by real time:
+
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 < the timestamp of T2.
+- Transaction T1 is ordered before Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is write-only, and
+  T2 is read-only.
+- Transaction T1 can be arbitrarily ordered with Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is
+  read-only, and T2 is read-only.
+- Transaction T1 can be arbitrarily ordered with Transaction T2 if the timestamp of T1 == the timestamp of T2, T1 is
+  write-only, and T2 is write-only.
```

Comment:
It feels incorrect to say that these transactions can be arbitrarily ordered right after saying that their order MUST be constrained by real time. However practically speaking the order of multiple consecutive writes don't actually matter if there's no reads in between them to view some intermediate state. We can practically view all these writes as a single write event, which is why it's ok to give them all the same timestamp. Similarly, the order of multiple consecutive reads don't matter if there's no write in between them to change the values read.

Reply:
I think the important thing to note is that they can be _arbitrary_, but once an order is chosen, it can NEVER be violated by any query in the future. That is, you can never observe state that can only happen if `T1` happens before `T2`, but we had arbitrarily chosen `T2` before `T1` in some other scenario

Reply:
Wouldn't they still be ordered by "timestamp assignment order"? E.g. let's imagine the oracle produces `(Timestamp, usize)` where the usize counts up. That is now the order. We strip off the `usize` when we submit reads and writes.